### PR TITLE
V0.2 explorer

### PR DESCRIPTION
**Pull request by ysqyang**
# Description

Redesign of exploration abstraction  

## Linked issue(s)/Pull request(s)

- [Issue 171](https://github.com/microsoft/maro/issues/171)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [x] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [x] Linux
- Python version:
  - [ ] 3.6
  - [x] 3.7
- Key information snapshot(s):
  
## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [x] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl/agent/abs_agent_manager.py [hunk 1]
```diff
@@ -82,10 +87,13 @@ def train(self, *args, **kwargs):
...
-    @property
-    def name(self):
-        """Agent manager's name."""
-        return self._name
+    def load_exploration_params(self, exploration_params):
+        if isinstance(exploration_params, dict) and exploration_params.keys() <= self.agent_dict.keys():
+            for agent_id, params in exploration_params.items():
+                self.agent_dict[agent_id].load_exploration_params(params)
+        else:
+            for agent in self.agent_dict.values():
```
Comment (ArthurJiang) [inline maro/rl/agent/abs_agent_manager.py:95]: add comments for per agent exploration

File: maro/rl/agent/simple_agent_manager.py [hunk 2]
```diff
@@ -42,20 +42,18 @@ def __init__(
...
-            "state": model_state,
-            "action": model_action,
+            "state": state,
+            "action": action,
             "reward": None,
             "agent_id": agent_id,
             "event": decision_event
         }
-        return self._action_shaper(model_action, decision_event, snapshot_list)
+        return self._action_shaper(action, decision_event, snapshot_list)
```
Comment (ArthurJiang) [inline maro/rl/agent/simple_agent_manager.py:56]: revert

File: maro/rl/exploration/epsilon_greedy_explorer.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    Args:
+        num_actions (int): Number of all possible actions.
+    """
+    def __init__(self, num_actions: int):
+        super().__init__()
+        self._num_actions = num_actions
+        self._epsilon = None
+
+    def __call__(self, action):
```
Comment (ArthurJiang) [inline maro/rl/exploration/epsilon_greedy_explorer.py:22]: action_index: int

Comment (ArthurJiang) [reply maro/rl/exploration/epsilon_greedy_explorer.py:22]: assert()

File: maro/rl/exploration/epsilon_greedy_explorer.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+class EpsilonGreedyExplorer(AbsExplorer):
+    """Epsilon greedy explorer for discrete action spaces.
+
+    Args:
+        num_actions (int): Number of all possible actions.
+    """
+    def __init__(self, num_actions: int):
+        super().__init__()
+        self._num_actions = num_actions
+        self._epsilon = None
```
Comment (ArthurJiang) [inline maro/rl/exploration/epsilon_greedy_explorer.py:20]: __init__(self, ..., epsilon: float = 0.0)
update/reschedule_explorer()

File: maro/rl/exploration/epsilon_schedule.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
```
Comment (ArthurJiang) [inline maro/rl/exploration/epsilon_schedule.py:2]: e_greedy_explorer, e_greedy_sheduler

File: maro/rl/exploration/noise_explorer.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    @abstractmethod
+    def __call__(self, action):
+        return NotImplementedError
+
+
+class UniformNoiseExplorer(NoiseExplorer):
+    """Explorer that adds a random noise to a model-generated action sampled from a uniform distribution."""
+    def __init__(
+        self,
```
Comment (ArthurJiang) [inline maro/rl/exploration/noise_explorer.py:39]: delta_noise: float = 0

Comment (ArthurJiang) [reply maro/rl/exploration/noise_explorer.py:39]: +_noise: Union[float, np.ndarray]?
-_noise: Union[float, np.npdarry]?

File: maro/rl/exploration/noise_explorer.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        action_dim: int,
+        min_action: Union[float, np.ndarray] = None,
+        max_action: Union[float, np.ndarray] = None
+    ):
+        super().__init__(action_dim, min_action, max_action)
+        self._noise_scale = None
+
+    def load_exploration_params(self, *, noise_scale: float):
+        self._noise_scale = noise_scale
```
Comment (ArthurJiang) [inline maro/rl/exploration/noise_explorer.py:75]: doc: mean: 0, std: noise_scale

Comment (ArthurJiang) [reply maro/rl/exploration/noise_explorer.py:75]: scale is relative or absolute value?

Comment (ArthurJiang) [reply maro/rl/exploration/noise_explorer.py:75]: base on current action relative noise

File: maro/rl/exploration/epsilon_greedy_schedule.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
```
Comment (ArthurJiang) [inline maro/rl/exploration/epsilon_greedy_schedule.py:3]: xx_scheduler.py

File: maro/rl/learner/simple_learner.py [hunk 9]
```diff
@@ -18,52 +17,29 @@ class SimpleLearner(AbsLearner):
...
-            Defaults to None.
         logger (Logger): Used to log important messages.
     """
     def __init__(
         self,
-        trainable_agents: SimpleAgentManager,
+        agent_manager: SimpleAgentManager,
         actor: Union[SimpleActor, ActorProxy],
-        explorer: Union[dict, AbsExplorer] = None,
         logger: Logger = DummyLogger()
```
Comment (ArthurJiang) [inline maro/rl/learner/simple_learner.py:29]: max_ep, scheduler_options: {class, paramters}

File: maro/rl/exploration/noise_explorer.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+class GaussianNoiseExplorer(NoiseExplorer):
+    """Explorer that adds a random noise to a model-generated action sampled from a Gaussian distribution."""
+    def __init__(
+        self,
+        action_dim: int,
+        min_action: Union[float, np.ndarray] = None,
+        max_action: Union[float, np.ndarray] = None,
+        noise_mean: Union[float, np.ndarray] = .0,
+        noise_stddev: Union[float, np.ndarray] = .0,
+        is_relative_stddev: bool = False
```
Comment (ArthurJiang) [inline maro/rl/exploration/noise_explorer.py:69]: is_relative_noise
if True:
  action += action * noise
else:
  action += noise

File: examples/cim/dqn/components/agent.py [hunk 11]
```diff
@@ -14,9 +14,17 @@ class CIMAgent(AbsAgent):
...
     """
-    def __init__(self, name, algorithm, experience_pool: ColumnBasedStore, min_experiences_to_train,
-                 num_batches, batch_size):
-        super().__init__(name, algorithm, experience_pool)
+    def __init__(
+        self,
+        name: str,
+        algorithm,
+        explorer: EpsilonGreedyExplorer,
+        experience_pool: ColumnBasedStore,
```
Comment (ArthurJiang) [inline examples/cim/dqn/components/agent.py:22]: add doc string for the parameters

File: examples/cim/dqn/components/agent_manager.py [hunk 12]
```diff
@@ -41,9 +42,7 @@ def create_dqn_agents(agent_id_list, config):
 
         experience_pool = ColumnBasedStore(**config.experience_pool)
         agent_dict[agent_id] = CIMAgent(
-            name=agent_id,
-            algorithm=algorithm,
-            experience_pool=experience_pool,
+            agent_id, algorithm, EpsilonGreedyExplorer(num_actions), experience_pool,
```
Comment (ArthurJiang) [inline examples/cim/dqn/components/agent_manager.py:45]: I don't think it's enough for initializing an explorer instance.

File: examples/cim/dqn/config.yml [hunk 13]
```diff
@@ -2,8 +2,15 @@ env:
   scenario: "cim"
   topology: "toy.4p_ssdd_l0.0"
   durations: 1120
-general:
+main_loop:
```
Comment (ArthurJiang) [inline examples/cim/dqn/config.yml:5]: learner_policy

Comment (ArthurJiang) [reply examples/cim/dqn/config.yml:5]: we need to refactor this part:
learner_policy
- main_loop
-- max_episode
-- early_stopping
- agent
-- rl_algorithms
--- exploration
-- learning_model
--- feature_engineer
--- learning_parameter

Comment (ArthurJiang) [reply examples/cim/dqn/config.yml:5]: need an offline discussion

File: examples/cim/dqn/dist_actor.py [hunk 14]
```diff
@@ -47,7 +47,7 @@ def launch(config, distributed_config):
         "max_retries": 15
     }
     actor_worker = ActorWorker(
-        local_actor=SimpleActor(env=env, inference_agents=agent_manager),
+        local_actor=SimpleActor(env, agent_manager),
```
Comment (ArthurJiang) [inline examples/cim/dqn/dist_actor.py:50]: use key=v, style

File: maro/rl/scheduling/exploration_parameter_generator.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+import numpy as np
+
+
+class StaticExplorationParameterGenerator(object):
+    """Exploration parameter generator based on a pre-defined schedule.
+
+    Args:
+        max_ep (int): Maximum number of episodes to run.
+    """
+    def __init__(self, max_ep: int, **config):
```
Comment (ArthurJiang) [inline maro/rl/scheduling/exploration_parameter_generator.py:16]: should not use config, nobody knows what's that

**Review by wesley-stone — APPROVED** [review]

**Review by wesley-stone — APPROVED** [review]

**Pull request merged**